xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authentication/xwiki-platform-security-authentication-default/src/main/java/org/xwiki/security/authentication/internal/ResetPasswordMailSender.java

```diff
@@ -0,0 +1,151 @@
+
+/**
+ * Component dedicated to send reset password information by email.
+ *
+ * @version $Id$
+ * @since 13.1RC1
+ */
+@Component(roles = ResetPasswordMailSender.class)
+@Singleton
+public class ResetPasswordMailSender
+{
+    private static final LocalDocumentReference RESET_PASSWORD_MAIL_TEMPLATE_REFERENCE =
+        new LocalDocumentReference("XWiki", "ResetPasswordMailContent");
+
+    @Inject
+    private MailSenderConfiguration mailSenderConfiguration;
+
+    @Inject
+    private DocumentReferenceResolver<EntityReference> documentReferenceResolver;
+
+    @Inject
+    @Named("template")
+    private MimeMessageFactory<MimeMessage> mimeMessageFactory;
+
+    @Inject
+    private MailSender mailSender;
+
+    @Inject
+    @Named("context")
+    private ComponentManager componentManager;
+
+    @Inject
+    private SessionFactory sessionFactory;
+
+    @Inject
+    private ContextualLocalizationManager localizationManager;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    /**
+     * Send the reset password information by email.
+     *
+     * @param username the name of the user for which to reset the password.
+     * @param email the email of the user for which to reset the password.
+     * @param resetPasswordURL the URL to use for resetting the password.
+     * @throws ResetPasswordException in case of error when preparing or sending the email.
+     */
+    public void sendResetPasswordEmail(String username, InternetAddress email, URL resetPasswordURL) throws
+        ResetPasswordException
+    {
+        XWikiContext context = this.contextProvider.get();
+        String fromAddress = this.mailSenderConfiguration.getFromAddress();
+        if (StringUtils.isEmpty(fromAddress)) {
+            fromAddress = "no-reply@" + context.getRequest().getServerName();
+        }
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("from", fromAddress);
+        parameters.put("to", email);
+        parameters.put("language", context.getLocale());
+        parameters.put("type", "Reset Password");
+        Map<String, String> velocityVariables = new HashMap<>();
+        velocityVariables.put("userName", username);
+        velocityVariables.put("passwordResetURL", resetPasswordURL.toExternalForm());
+        parameters.put("velocityVariables", velocityVariables);
+
+        String localizedError =
+            this.localizationManager.getTranslationPlain("xe.admin.passwordReset.error.emailFailed");
+
+        MimeMessage message;
+        try {
+            message =
+                this.mimeMessageFactory.createMessage(
+                    this.documentReferenceResolver.resolve(RESET_PASSWORD_MAIL_TEMPLATE_REFERENCE), parameters);
+        } catch (MessagingException e) {
+            throw new ResetPasswordException(localizedError, e);
+        }
+        try {
+            MailListener mailListener = this.componentManager.getInstance(MailListener.class, "database");
```

Comment:
why not inject it and avoid the try/catch?

Reply:
Because my MailSender component is a singleton, and this component should be instantiated each time I call the method.

Reply:
For that you can use:

```
@Inject
@Named("database")
private Provider<MailListener> databaseMailListenerProvider;
```

Reply:
Ah indeed I missed that option.